### ISSUE TITLE
Support for intercepting all requests sent to HttpClient

### ISSUE THREAD

**Issue by stephenthrelfall** [justeat/httpclient-interception#20]
Would it be possible to easily implement intercepting all requests (i.e. not specifying a host) for the purposes of then doing some sort of assertions on the request. 

For example if I am unit testing a class which uses HttpClient and I don't care about the host or the verb that is being sent to (this is covered by other tests) I simply want to test that the correct headers are being applied.

**Comment by martincostello** [comment]
I've just refreshed myself with the code, and it looks like _part_ of what you want is already there, but not all of it.

We do have an `OnSend` property on `HttpClientInterceptorOptions` that you can use to inspect every request as they flow through:

https://github.com/justeat/httpclient-interception/blob/204e16d7a5733cff68274564a60424852c2f390d/src/HttpClientInterception/HttpClientInterceptorOptions.cs#L57-L60

But ultimately that still calls through into the code to find a matching response, which expects to find a registration, and if not, calls through to the inner handler (unless `ThrowOnMissingRegistration` is set to `true`), which means you still need to register _something_ to match the request.

I guess what you're kind of after is a way to register a "catch-all" response for anything that doesn't match something more specific?

**Comment by martincostello** [comment]
Something like this maybe?

```csharp
var options = new HttpClientInterceptorOptions();

options.OnMissingRegistration = (request) =>
{
    return Task.FromResult(new HttpResponseMessage());
};
```

**Comment by stephenthrelfall** [comment]
That's exactly it. Something whereby I can register only on the fields I currently care about and ignores the rest. I'm happy to put together a pr, just wondered what your thoughts were on feasibility and whether it makes sense.

On 1 Mar 2018 18:29, "Martin Costello" <notifications@github.com<mailto:notifications@github.com>> wrote:

I've just refreshed myself with the code, and it looks like part of what you want is already there, but not all of it.

We do have an OnSend property on HttpClientInterceptorOptions that you can use to inspect every request as they flow through:

https://github.com/justeat/httpclient-interception/blob/204e16d7a5733cff68274564a60424852c2f390d/src/HttpClientInterception/HttpClientInterceptorOptions.cs#L57-L60

But ultimately that still calls through into the code to find a matching response, which expects to find a registration, and if not, calls through to the inner handler (unless ThrowOnMissingRegistration is set to true), which means you still need to register something to match the request.

I guess what you're kind of after is a way to register a "catch-all" response for anything that doesn't match something more specific?

**Comment by martincostello** [comment]
I like the idea of an optional “use this if nothing better” - it’s a use case I hadn’t really considered when coming up with the initial design.

I think something fairly “generic” that supports being asynchronous, for example if you want to read from a file to get the default body, should be good enough to accomplish your use case.

I’m keen to not have too much pattern matching etc. being built in, as that’s got the potential to lead to bloat trying to cater for all the things people might want to do in their test scenarios.

Instead, I think hooks like this allow enough extensibility to “bring your own implementation” without the library getting too opinionated.

Feel free to submit a PR!

**Comment by stephenthrelfall** [comment]
I had envisaged it working more like this:

```csharp
var builder = new HttpRequestInterceptionBuilder()
                    .ForHttp()
                    .ForPath("orders")                    
                    .WithContent("{}");
```
Would match any request with orders as the path and any query string. 

```csharp
var builder = new HttpRequestInterceptionBuilder()
                    .ForHttp()
                    .ForPath("orders") 
                    .ForQuery($"id=12")                   
                    .WithContent("{}");
```
Would match any request with orders as the path and that query string.

Maybe that is what you were referring to when you said too much pattern matching? Maybe this could be a separate kind of builder?

**Comment by martincostello** [comment]
The problem with that approach is that that would break the current behaviour, where an unspecified host implies `localhost` and no query string implies no query string at all.

Ignoring query strings from matching behaviour was added by #10. We could maybe add in an "any host" matching behaviour in the same vein?

However then I'd be worried that in tests for an application that had more than one external dependency that users could mistakenly miss issues where their tests were hitting the "wrong" host name and they wouldn't notice because it now wouldn't matter.

I wonder if there's a way we could combine the two approaches somehow? Have a general "catch all" hook to allow you to provide you the opportunity to provide an arbitrary response for an arbitrary request that hasn't matched already, but then have a simple builder available that you can use to build a response easily after the "bring-your-own" condition has been satisfied?

**Comment by stephenthrelfall** [comment]
I think requiring users to specifically call the AnyHost() method to get that kind of behaviour would mitigate the risk of it being done mistakenly and is quite explicit in the test, which I definitely like.

**Comment by martincostello** [comment]
Would you be able to share a code example of roughly what you're trying to achieve in your specific user case?

I'm curious as to what the bigger picture is that's raising the need to use this library in-lieu of a simpler unit test of a `HttpMessageHandler` implementation coupled with a mocking framework.

That aside, I like the idea of something like `AnyHost()` being the way to enable the behaviour if we implemented it, though off the top of my head I'm not sure how we'd implement it under the hood as a lot of the current behaviour leverages off the back of `UriBuilder`.

**Comment by stephenthrelfall** [comment]
I used the library to replace httpclient in my unit, integration and functional tests. I guess I favoured this approach over rolling my own as it provides a common language used throughout my tests for mocking the httpclient and it provides almost all the functionality I required already.

**Comment by martincostello** [comment]
Right OK, so more for consistency of approach across your different levels of test?

**Comment by stephenthrelfall** [comment]
It's also about clarity. I think this example using your library: 
```csharp
                // Setup an expected response
                var builder = new HttpRequestInterceptionBuilder()
                    .ForHttp()
                    .ForUri(connectionDetails.ConnectionUri)
                    .ForPath(_getRestaurantPath)
                    .ForQuery($"q={outcode}")
                    .WithStatus(HttpStatusCode.OK)
                    .WithContent(testJson);
```
Conveys it's intentions much more clearly than something like this:
```csharp
            var messageHandler = new Mock<MockableMessageHandler>();
            var httpClient = new HttpClient(messageHandler.Object);

            messageHandler
                .Setup(m => m.DoSendAsync(It.IsAny<HttpRequestMessage>()))
                .ReturnsAsync(new HttpResponseMessage()
            {
                Content = new StringContent(testJson)
            });
```

**Comment by martincostello** [comment]
I'll have a tinker around in a branch later this afternoon and see if some inspiration strikes.

**Comment by martincostello** [comment]
Have raised a PR for further discussion and review based on the ideas here in #21.

**Comment by martincostello** [comment]
Resolved by #21.
